crates/sui-core/src/authority/backpressure.rs

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+
+use mysten_metrics::monitored_scope;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use tokio::sync::watch;
+use tracing::{debug, info};
+
+use crate::checkpoints::CheckpointStore;
+
+#[derive(Debug, Default, Copy, Clone)]
+struct Watermarks {
+    executed: CheckpointSequenceNumber,
+    certified: CheckpointSequenceNumber,
+}
+
+impl Watermarks {
+    // we can only permit backpressure if the certified checkpoint is ahead of the executed
+    // checkpoint. Otherwise, backpressure might prevent construction of the next checkpoint,
+    // because it could stop consensus commits from being processed.
+    fn is_backpressure_suppressed(&self) -> bool {
+        self.certified <= self.executed
+    }
+}
+
+pub struct BackpressureManager {
+    // holds the executed and certified checkpoint watermarks.
+    // builder can run ahead of executor, or vice versa, so these watermarks
+    // do not have a strict ordering.
+    //
+    // We want to engage backpressure whenever builder is running ahead of the executor.
```

Comment:
this comment was totally out of date so i'm re-writing it. (an early version of this code used executed and built watermarks instead of executed and certified)